Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,192 @@
+        Connect-AzureADTenant -tenantId $tenantId
+
+        Write-DosMessage -Level "Information" -Message "Retrieving group $($groupSID) from Azure AD Tenant $($tenantId)..."
+        $azureADGroups = Get-AzureADGroup -Filter "onPremisesSecurityIdentifier eq '$($groupSID)'"
+
+        # disconnect from Azure AD
+        Disconnect-AzureAD
+    }
+    catch {
+        Write-DosMessage -Level "Fatal" -Message "Error while attempting to retrieve Azure AD group $($groupSID): $($_.Exception)"
+    }
+
+    if ($azureADGroups.Count -eq 1) {
+        return $azureADGroups[0]
+    }
+    elseif ($azureADGroups.Count -eq 0) {
+        $errorMsg = "No match found for SID $($groupSID) in Azure AD."
+        Write-DosMessage -Level "Error" -Message $errorMsg
+        throw $errorMsg
+    }
+    else {
+        $errorMsg = "Multiple matches found for group SID $($groupSID)."
+        Write-DosMessage -Level "Error" -Message $errorMsg
+        throw $errorMsg
+    }
+}
+
+function Move-ActiveDirectoryGroupsToAzureAD {
+     param(
+        [Parameter(Mandatory=$true)]
+        [string] $connString
+    )
+
+    Write-DosMessage -Level "Information" -Message "Migrating AD groups to Azure AD..."
+
+    $allowedTenantIds = Get-AzureADTenants -installConfigPath $installConfigPath
+    if($null -eq $allowedTenantIds -or $allowedTenantIds.Count -eq 0) {
+        Write-DosMessage -Level "Error" -Message  "No tenants were found in the install.config"
+        throw
+    }
+
+    # retrieve all groups from Auth DB    
+    try {
+        $results = Get-NonMigratedActiveDirectoryGroups -connectionString $connString
+    }
+    catch {
+        Write-DosMessage -Level "Fatal" -Message "An error occurred while executing the command to retrieve non-migrated AD groups. Connection String: $($connString). Error $($_.Exception)"
+    }
+
+    if ($results.Count -eq 0) {
+        Write-DosMessage -Level "Information" -Message "No groups found to migrate."
+    }
+
+    foreach ($group in $results) {
+        # query AD for SID
+        $adGroupSID = ""
+        try {
+            $authGroupNameParts = $group.Name.Split('\')
+            $adGroupName = ""
+            if ($authGroupNameParts.Count -eq 1) {
+                $adGroupName = $authGroupNameParts[0]
+            }
+            else {
+                $adGroupName = $authGroupNameParts[1]
+            }
+            $adGroup = Get-ADGroup -Identity $adGroupName
+            $adGroupSID = $adGroup.SID.Value
+        }
+        catch {
+            Write-DosMessage -Level "Fatal" -Message "An error occurred while retrieving AD Group $($group.Name) from Active Directory. Error $($_.Exception)"
+            continue
+        }
+
+        foreach ($tenantId in $allowedTenantIds) {
+            # query Azure AD to get external ID
+            try {
+                $azureADGroup = Get-AzureADGroupBySID -tenantId $tenantId -groupSID $adGroupSID
+            }
+            catch {
+                continue
```

Comment:
Log here even if this is an OK exception to help with diagnostics/debug.

Reply:
I log each of the paths in `Get-AzureADGroupBySID`. This is just catching any potential exceptions and attempting to find the group in the next tenant.